blockchain/validate.go

```diff
@@ -562,15 +569,25 @@ func checkBlockSanity(block *btcutil.Block, powLimit *big.Int, timeSource Median
 		// We could potentially overflow the accumulator so check for
 		// overflow.
 		lastSigOps := totalSigOps
-		totalSigOps += CountSigOps(tx)
-		if totalSigOps < lastSigOps || totalSigOps > MaxSigOpsPerBlock {
+		totalSigOps += (CountSigOps(tx) * WitnessScaleFactor)
+		if totalSigOps < lastSigOps || totalSigOps > MaxBlockSigOpsCost {
 			str := fmt.Sprintf("block contains too many signature "+
 				"operations - got %v, max %v", totalSigOps,
-				MaxSigOpsPerBlock)
+				MaxBlockSigOpsCost)
 			return ruleError(ErrTooManySigOps, str)
 		}
 	}
 
+	// Once the witness commitment, witness nonce, and sig op cost have
+	// been validated, we can finally assert that the block's weight
+	// doesn't exceed the current consensus paramter.
+	blockWeight := GetBlockWeight(block)
```

Comment:
I understand what you're saying, and I agree with the fact that `GetBlockWeight` itself does not need to rely on the activation state.  However, what I said is that _this particular check_ can't go in this function and I still believe that to be true.  I should've explained why though.

The reason for this is that you can't properly check the block weight without undesired side effects until you've determined whether or not the witness nonce and commitments are valid (when they're active).  Failure to do this means it would be possible to create an invalid witness that is full of data and would therefore cause the block weight to be larger than allowed and thus result in an invalid block.  That in and of itself doesn't sound like a problem, but if we consider the fact we're talking about the witness here, we have to keep in mind that the block hash won't change regardless of its contents.  This means that an attacker could create an invalid block that confuses the various caches which keep track of recently rejected blocks based on their hash and such.